### PR TITLE
ARROW-8742: [C++][Python] Add GRPC Mutual TLS for clients and server

### PR DESCRIPTION
**Pull request by erinleeryan**
Our team would like to implement mutual tls authentication for a pyarrow client, hence modifications to the C++ and Python implementations to at a GrpcMTls scheme for connections.

### CODE REVIEW

File: cpp/src/arrow/flight/client.cc [hunk 1]
```diff
@@ -531,14 +531,19 @@ class FlightClient::FlightClientImpl {
...
+    if (scheme == kSchemeGrpc || scheme == kSchemeGrpcTcp || scheme == kSchemeGrpcTls || scheme == kSchemeGrpcMTls) {
       grpc_uri << location.uri_->host() << ":" << location.uri_->port_text();
 
-      if (scheme == "grpc+tls") {
+      if ((scheme == "grpc+tls") || (scheme == "grpc+mtls") {
         grpc::SslCredentialsOptions ssl_options;
         if (!options.tls_root_certs.empty()) {
           ssl_options.pem_root_certs = options.tls_root_certs;
         }
+        if (scheme  == 'grpc+mtls'){
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client.cc:542]: Is a whole new scheme necessary? I'd say that so long as the client has provided a list of client cert/key pairs, then mTLS should just be enabled, no?

Comment (lidavidm) [reply cpp/src/arrow/flight/client.cc:542]: Also, this doesn't build and needs to be formatted - the developer guide has instructions: https://arrow.apache.org/docs/developers/cpp/index.html

File: cpp/src/arrow/flight/server.cc [hunk 2]
```diff
@@ -665,16 +665,20 @@ Status FlightServerBase::Init(const FlightServerOptions& options) {
...
     address << location.uri_->host() << ':' << location.uri_->port_text();
 
     std::shared_ptr<grpc::ServerCredentials> creds;
-    if (scheme == kSchemeGrpcTls) {
+    if ((scheme == kSchemeGrpcTls) || (scheme == kSchemeGrpcMTls)) {
       grpc::SslServerCredentialsOptions ssl_options;
       for (const auto& pair : options.tls_certificates) {
         ssl_options.pem_key_cert_pairs.push_back({pair.pem_key, pair.pem_cert});
       }
+      if (scheme == kSchemeGrpcMTls) {
```
Comment (lidavidm) [inline cpp/src/arrow/flight/server.cc:678]: I'd say instead of using a new scheme, just add a flag to `FlightServerOptions`.

**Review by lidavidm — CHANGES_REQUESTED** [review]

**Review by lidavidm — CHANGES_REQUESTED** [review]
Thanks for the fixes! There are some style things that need to be fixed, or I can fix them with access to this PR; also, it seems there are some changes in _flight.pyx that are missing.

File: cpp/src/arrow/flight/server.h [hunk 3]
```diff
@@ -25,6 +25,7 @@
 #include <string>
 #include <utility>
 #include <vector>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (erinleeryan) [reply cpp/src/arrow/flight/server.h]: That was a remnant of me trying to use boolalpha for the the client verification flag, it's now removed.

File: cpp/src/arrow/flight/server.cc [hunk 4]
```diff
@@ -675,6 +675,10 @@ Status FlightServerBase::Init(const FlightServerOptions& options) {
       for (const auto& pair : options.tls_certificates) {
         ssl_options.pem_key_cert_pairs.push_back({pair.pem_key, pair.pem_cert});
       }
+      if (!options.verify_client.empty()) {
+        ssl_options.client_certificate_request =
```
[the remark this answers is not in the document]
Comment (erinleeryan) [reply cpp/src/arrow/flight/server.cc:679]: I double checked, this should allow edits by maintainers as it currently stands, this is certainly one spot where trying to go from Python-land's loose sloppiness as a coder to C++ is perhaps suboptimal.

**Comment by lidavidm** [comment]
I've reformatted the code and added a test to make sure mTLS actually works, let's see how it fares in CI.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Merged, thank you for implementing this!

**Pull request closed without merging**